File: 2678-design-graph-with-shortest-path-calculator/design-graph-with-shortest-path-calculator.cpp
class Graph {
public:
    typedef pair<int, int> P;
    unordered_map<int, vector<P>> adj;
    int N;
    priority_queue<P, vector<P>, greater<P>> pq;
    Graph(int n, vector<vector<int>>& edges) {
        N = n;
        for (vector<int>& edge : edges) {
            int u = edge[0];
            int v = edge[1];
            int cost = edge[2];
            adj[u].push_back({v, cost});
        }
    }

    void addEdge(vector<int> edge) {
        int u = edge[0];
        int v = edge[1];
        int cost = edge[2];
        adj[u].push_back({v, cost});
    }

    int shortestPath(int node1, int node2) {
        vector<int>result(N,INT_MAX);
        result[node1]=0;
        pq.push({0,node1});

        while(!pq.empty()){
            int currDist = pq.top().first;
            int currNode = pq.top().second;
            pq.pop();

            for(auto next:adj[currNode]){
                int nextNode = next.first;
                int nextDist = next.second;
                if(currDist+nextDist<result[nextNode]){
                    result[nextNode]= currDist+nextDist;
                    pq.push({result[nextNode],nextNode});
                }
            }

        }

        return result[node2]==INT_MAX ? -1 : result[node2];
    }
};

/**
 * Your Graph object will be instantiated and called as such:
 * Graph* obj = new Graph(n, edges);
 * obj->addEdge(edge);
 * int param_2 = obj->shortestPath(node1,node2);
 */

File: 2678-design-graph-with-shortest-path-calculator/design-graph-with-shortest-path-calculator.py
import heapq
from collections import defaultdict
class Graph:

    def __init__(self, n: int, edges: List[List[int]]):
        self.N = n
        self.adj = defaultdict(list)
        for u,v,dist in edges:
            self.adj[u].append((v,dist))


    def addEdge(self, edge: List[int]) -> None:
        self.adj[edge[0]].append((edge[1],edge[2]))
        

    def shortestPath(self, node1: int, node2: int) -> int:
        result = [float('inf')]*self.N
        result[node1]=0 
        pq = [(0,node1)]
        while pq:
            curr_dist , curr_node = heapq.heappop(pq)

            if(curr_dist > result[curr_node]):
                continue

            for next_node , next_dist in self.adj[curr_node]:
                if(curr_dist+next_dist<result[next_node]):
                    result[next_node]=curr_dist+next_dist
                    heapq.heappush(pq,(result[next_node],next_node))

        return result[node2] if result[node2]!=float('inf') else -1            
        


# Your Graph object will be instantiated and called as such:
# obj = Graph(n, edges)
# obj.addEdge(edge)
# param_2 = obj.shortestPath(node1,node2)